pkg/agent/proxy/proxier.go

```diff
@@ -241,10 +267,77 @@ func smallSliceDifference(s1, s2 []string) []string {
 	return diff
 }
 
+func (p *proxier) installNodePortService(groupID binding.GroupIDType, svcPort uint16, protocol binding.Protocol, affinityTimeout uint16, nodeLocalExternal bool) error {
+	if err := p.ofClient.InstallServiceFlows(groupID, agentconfig.DummyNodePortSvcIP, svcPort, protocol, affinityTimeout); err != nil {
+		return fmt.Errorf("failed to install Service NodePort load balancing flows: %w", err)
+	}
+	if err := p.ofClient.InstallServiceClassifierFlow(corev1.ServiceTypeNodePort, agentconfig.DummyNodePortSvcIP, svcPort, protocol, nodeLocalExternal); err != nil {
+		return fmt.Errorf("failed to install Service NodePort classifying flows: %w", err)
+	}
+	if err := p.routeClient.AddNodePort(p.nodePortIPMap, svcPort, protocol, p.isIPv6); err != nil {
+		return fmt.Errorf("failed to install Service NodePort traffic redirecting flows: %w", err)
+	}
+	return nil
+}
+
+func (p *proxier) uninstallNodePortService(svcPort uint16, protocol binding.Protocol) error {
+	if err := p.ofClient.UninstallServiceFlows(agentconfig.DummyNodePortSvcIP, svcPort, protocol); err != nil {
+		return fmt.Errorf("failed to remove Service NodePort NodePort load balancing flows: %w", err)
+	}
+	if err := p.ofClient.UninstallServiceClassifierFlow(agentconfig.DummyNodePortSvcIP, svcPort, protocol); err != nil {
+		return fmt.Errorf("failed to remove Service NodePort classifying flows: %w", err)
+	}
+	if err := p.routeClient.DeleteNodePort(p.nodePortIPMap, svcPort, protocol, p.isIPv6); err != nil {
+		return fmt.Errorf("failed to remove Service NodePort traffic redirecting flows: %w", err)
+	}
+	return nil
+}
+
+func (p *proxier) installLoadBalancerService(groupID binding.GroupIDType, loadBalancerIPStrings []string,
+	svcPort uint16, protocol binding.Protocol, affinityTimeout uint16, nodeLocalExternal bool) error {
+	for _, ingress := range loadBalancerIPStrings {
+		if ingress != "" {
+			if err := p.ofClient.InstallServiceFlows(groupID, net.ParseIP(ingress), svcPort, protocol, affinityTimeout); err != nil {
+				return fmt.Errorf("failed to install Service LoadBalancer load balancing flows: %w", err)
+			}
+			if err := p.ofClient.InstallLoadBalancerServiceFromOutsideFlows(net.ParseIP(ingress), svcPort, protocol); err != nil {
+				return fmt.Errorf("failed to install Service LoadBalancer flows: %w", err)
+			}
+			if err := p.ofClient.InstallServiceClassifierFlow(corev1.ServiceTypeLoadBalancer, net.ParseIP(ingress), svcPort, protocol, nodeLocalExternal); err != nil {
+				return fmt.Errorf("failed to install Service LoadBalancer classifying flows: %w", err)
+			}
+		}
+	}
+	if err := p.routeClient.AddLoadBalancer(loadBalancerIPStrings, p.isIPv6); err != nil {
```

Comment:
Can loadBalancerIPStrings include ipv4 and ipv6 at same time in dual-stack environment? Can AddLoadBalancer handle it correctly?

Reply:
I missed this case. I'll rewrite this function. Judging every ingress IP is ok.